presto-main/src/main/java/com/facebook/presto/server/security/JsonWebTokenAuthenticator.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server.security;
+
+import com.facebook.presto.spi.security.BasicPrincipal;
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.Jws;
+import io.jsonwebtoken.JwtParser;
+import io.jsonwebtoken.Jwts;
+
+import javax.inject.Inject;
+import javax.servlet.http.HttpServletRequest;
+
+import java.security.Principal;
+
+import static com.google.common.base.Strings.nullToEmpty;
+import static com.google.common.net.HttpHeaders.AUTHORIZATION;
+
+public class JsonWebTokenAuthenticator
+        implements Authenticator
+{
+    private final JwtParser jwtParser;
+
+    @Inject
+    public JsonWebTokenAuthenticator(JsonWebTokenConfig config)
+    {
+        JwtParser jwtParser = Jwts.parser()
+                .setSigningKey(config.getKey());
+        if (config.getRequiredIssuer() != null) {
+            jwtParser.requireIssuer(config.getRequiredIssuer());
+        }
+        if (config.getRequiredAudience() != null) {
+            jwtParser.requireAudience(config.getRequiredAudience());
+        }
+        this.jwtParser = jwtParser;
+    }
+
+    @Override
+    public Principal authenticate(HttpServletRequest request)
+            throws AuthenticationException
+    {
+        String header = nullToEmpty(request.getHeader(AUTHORIZATION));
+
+        int space = header.indexOf(' ');
+        if ((space < 0) || !header.substring(0, space).equalsIgnoreCase("bearer")) {
+            throw needAuthentication(null);
+        }
+        String token = header.substring(space + 1).trim();
+
+        try {
+            Jws<Claims> claimsJws = jwtParser.parseClaimsJws(token);
```

Comment:
This needs to catch `JwtException` (and maybe others) and call `needAuthentication()` with the message. A `RuntimeException` thrown from `authenticate()` will cause the server to return a 500, so that should only happen if there is an internal error, not due to an expired token or similar.

Reply:
Those errors should only happen if the token is not trusted or corrupted.  In other authenticators, they throw runtime when this happens.  I think we should have another mechanism to communicate these kinds of errors.